File: csharp/Ice/session/SessionFactoryI.cs
// Copyright (c) ZeroC, Inc.

using Demo;

internal class SessionFactoryI : SessionFactoryDisp_
{
    public override SessionPrx create(string name, Ice.Current current)
    {
        var session = new SessionI(name);
        var proxy = SessionPrxHelper.uncheckedCast(current.adapter.addWithUUID(session));

        // Remove endpoints to ensure that calls are collocated-only
        // This way, if we invoke on the proxy during shutdown, the invocation fails immediately
        // without attempting to establish any connection
        var collocProxy = SessionPrxHelper.uncheckedCast(proxy.ice_endpoints(new Ice.Endpoint[0]));

        current.con.setCloseCallback(_ =>
            {
                try
                {
                    collocProxy.destroy();
                    Console.Out.WriteLine("Cleaned up dead client.");
                }
                catch (Ice.LocalException)
                {
                    // The client already destroyed this session, or the server is shutting down
                }
            });
        return proxy;
    }

    public override void shutdown(Ice.Current current)
    {
        Console.Out.WriteLine("Shutting down...");
        current.adapter.getCommunicator().shutdown();
    }
}

File: java/Ice/session/SessionFactoryI.java
// Copyright (c) ZeroC, Inc.

import com.zeroc.demos.Ice.session.Demo.*;

class SessionFactoryI implements SessionFactory
{
    @Override
    public SessionPrx create(String name, com.zeroc.Ice.Current current)
    {
        SessionI session = new SessionI(name);
        SessionPrx proxy = SessionPrx.uncheckedCast(current.adapter.addWithUUID(session));

        //
        // Remove endpoints to ensure that calls are collocated-only
        // This way, if we invoke on the proxy during shutdown, the invocation fails immediately
        // without attempting to establish any connection
        //
        SessionPrx collocProxy = proxy.ice_endpoints(new com.zeroc.Ice.Endpoint[0]);

        current.con.setCloseCallback(con ->
                 {
                     try
                     {
                         collocProxy.destroy();
                         System.out.println("Cleaned up dead client.");
                     }
                     catch(com.zeroc.Ice.LocalException ex)
                     {
                         // The client already destroyed this session, or the server is shutting down
                     }
                 });

        return proxy;
    }

    @Override
    public void shutdown(com.zeroc.Ice.Current current)
    {
        System.out.println("Shutting down...");
        current.adapter.getCommunicator().shutdown();
    }
}
